src/CppAst/CppModelBuilder.cs

```diff
@@ -2139,6 +2158,13 @@ bool ConsumeIfTokenBeforeIs(ref CXSourceLocation loc, string token_str)
                                     ++parenCount;
 
                                 begin = GetPrevLocation(begin, 1);
+
+                                // We have reached the end of the source of trying to deal
+                                // with the potential of alignas, so we just break, which
+                                // will cause ConsumeIfTokenBeforeIs(ref begin, "alignas") to be false
+                                // and thus fall back to saveBegin which is the correct behavior
+                                if (begin.Equals(CXSourceLocation.Null))
```

Comment:
Are you sure that it is the only place it is required?

We should not assume that the code is well formed (you can have incomplete statement, or end-of-file happening in a loop), I'm worried that a few loops introduced are potentially not safe:

For example this one?

https://github.com/xoofx/CppAst/blob/3aa74ac39b19bae0b9c86530bf71236187b09e3c/src/CppAst/CppModelBuilder.cs#L2147-L2148

This one, if it never finds `(` or `)` in the loop?

https://github.com/xoofx/CppAst/blob/3aa74ac39b19bae0b9c86530bf71236187b09e3c/src/CppAst/CppModelBuilder.cs#L2195-L2202

Same for this one, if none of the `>`, `>>` or `<`, the variable `angleCount` would not be modified:

https://github.com/xoofx/CppAst/blob/3aa74ac39b19bae0b9c86530bf71236187b09e3c/src/CppAst/CppModelBuilder.cs#L2208-L2216

Reply:
So the first case I think is a legit issue because you could potentially parse a comment, and that could lead to the loop getting stuck and I have checked in a fix for that.

The second and third case are little bit more difficult because CppAST doesn't allow you to process invalid code, in that if there are any errors in the code we just skip processing completely. Both those cases are restricted to `CXCursor_TemplateTemplateParameter` which here again would require that CppAST allowed for the parsing of invalid C++ which is doesn't.

Reply:
> The second and third case are little bit more difficult because CppAST doesn't allow you to process invalid code

Oh right, I forgot that CppParser is indeed not proceeding on the translation unit if there are errors. good.